### ISSUE TITLE
Deduplication labels per tenant

### ISSUE THREAD

**Issue by brancz** [thanos-io/thanos#3871]
### Is your proposal related to a problem?

Different tenants may choose to label their ingesting Prometheus servers with different labels, Thanos should be flexible enough to allow for this.

### Describe the solution you'd like

I'm inclined to allow it per request in order to be able to have different label schemes even within one tenant, on the other hand the same per-tenant deduplication-label would be good to have at offline deduplication time, so a per-tenant configuration could be a good idea as well.

### Describe alternatives you've considered

Enforcing tenants to use the same labels for deduplication everywhere. This is not practical I believe as you can't necessarily enforce things like this, and even if it's likely that in the lifetime of running Thanos these might want to change.

@thanos-io/thanos-maintainers

**Comment by bwplotka** [comment]
Makes sense, thanks for raising. I think this screams for config object instead of flag though 🤔

**Comment by brancz** [comment]
Did I write flag anywhere?

**Comment by LeviHarrison** [comment]
@brancz I might take a swing at this if that's OK. I just have a few questions about the implementation:

* What I've gathered is that the proposal is to allow different deduplication labels per tenant so that when making a query encompassing multiple tenants, everything would be deduplicated properly. Is that correct?

* So there would be a configuration that looks something like this:

```yaml
replica_labels:
  - tenant1:
      - replicay
  - tenant2:
      - replica1
      - replicax
```

And that would be in a file, specified by a flag like `--query.replica-label-file=replica_labels.yaml`.

* You also mentioned having an option to do this per-request, which I think might get a little complicated regarding URL parameters, but maybe bending the existing `replicaLabels` parameter with something along the lines of this: 
```
?replicaLabels=tenant1=replicay&replicaLabels=tenant2=replica1
```
and then use a regex.

* I'm a little bit new to Thanos, how would you distinguish between tenants in the querier?

**Comment by brancz** [comment]
I think I would start with the config and not per request (it could still be done additionally later). I suspect that there will be various per tenant configurations eventually, so I think I would invert the config:

```
tenants:
- name: tenant1
  replica_labels:
  - replicay
- name: tenant2
  replica_labels:
  - replica1
  - replicax
```

**Comment by LeviHarrison** [comment]
Ok, so we use the value of the `THANOS_TENANT` header (or something like that) in a tenant label in the query and then deduplicate the results based on the replica labels specified in the config (if there are any specified). Does that seem right?

Sidenote: is there a specific label used to distinguish tenants, or will that have to be configured also?

**Comment by ron9295** [comment]
Hi everyone 👋
Would it make sense to implement this in the Querier component first?
That way, users who are running multiple Receivers with different tenant label schemes could still query them using a single multi-tenant Querier instance.
